File: plover_vm/macro/fast.py
"""Direct ISA semantics without micro-phases."""

from __future__ import annotations

from plover_vm.alu import alu8
from plover_vm.alu16 import add16, cmp16_u
from plover_vm.micro.normative import apply_add, apply_beq_compare, apply_cmp_flags
from plover_vm.macro.isa import (
    OP_ADD,
    OP_ADD_RR,
    OP_BCS,
    OP_BEQ,
    OP_CALL,
    OP_CMP,
    OP_HALT,
    OP_JMP,
    OP_LDA,
    OP_LDIO,
    OP_MOV,
    OP_RET,
    OP_STA,
    OP_STA16,
    OP_STIO,
    OP_WADD_RR,
    OP_WCMP16,
    OP_WMOV,
    WIDE_ABS16_OPS,
    WIDE_IMM16_OPS,
)
from plover_vm.macro.mmio import mmio_addr
from plover_vm.memory.bus import MemoryBus


class MacroFastPath:
    def __init__(self, bus: MemoryBus) -> None:
        self.bus = bus
        self.pc = 0
        self.regs = [0, 0, 0, 0]
        self.regs16 = [0, 0, 0, 0]
        self.flag_z = False
        self.flag_c = False
        self.halted = False
        self._ret_stack: list[int] = []

    def _read_byte(self, addr: int) -> int:
        return self.bus.read_cpu(addr) & 0xFF

    def fetch_decode_execute(self) -> None:
        if self.halted:
            return
        fa = self.bus.fetch_addr(self.pc)
        op = self._read_byte(fa)

        if op in WIDE_IMM16_OPS:
            imm16 = self._read_byte(fa + 1) | (self._read_byte(fa + 2) << 8)
            self.pc = (fa + 3) & 0xFFFF
            imm = imm16
        elif op in WIDE_ABS16_OPS:
            imm = self._read_byte(fa + 1) | (self._read_byte(fa + 2) << 8)
            self.pc = (fa + 3) & 0xFFFF
        elif op in (OP_RET, OP_HALT, OP_ADD_RR):
            imm = 0
            self.pc = (fa + 1) & 0xFFFF
        else:
            imm = self._read_byte(fa + 1)
            self.pc = (fa + 2) & 0xFFFF

        if op == OP_HALT:
            self.halted = True
        elif op == OP_ADD:
            self.regs, self.flag_z, self.flag_c = apply_add(self.regs, imm)
        elif op == OP_ADD_RR:
            r = alu8(self.regs[0], self.regs[1], 1)
            self.regs[2] = r.y
            self.flag_z = r.zero
            self.flag_c = r.cout
        elif op == OP_MOV:
            dst, src = (imm >> 4) & 3, imm & 3
            self.regs[dst] = self.regs[src] & 0xFF
        elif op == OP_CMP:
            self.flag_z, self.flag_c = apply_cmp_flags(self.regs, imm)
        elif op == OP_WADD_RR:
            r = add16(self.regs16[0], self.regs16[1])
            self.regs16[2] = r.y
            self.flag_z = r.zero
            self.flag_c = r.cout
        elif op == OP_WMOV:
            dst, src = (imm >> 4) & 3, imm & 3
            self.regs16[dst] = self.regs16[src] & 0xFFFF
        elif op == OP_WCMP16:
            r = cmp16_u(self.regs16[0], imm)
            self.flag_z = r.zero
            self.flag_c = r.cout
        elif op == OP_BCS:
            if self.flag_c:
                self.pc = imm & 0xFFFF
        elif op == OP_LDA:
            self.regs[0] = self._read_byte(imm)
        elif op == OP_STA:
            self.bus.write_cpu(imm, self.regs[0])
        elif op == OP_STA16:
            self.bus.write_cpu(imm, self.regs[0])
        elif op == OP_BEQ:
            self.flag_z, self.flag_c = apply_beq_compare(self.regs, imm)
            if self.flag_z:
                self.pc = imm & 0xFFFF
        elif op == OP_JMP:
            self.pc = imm & 0xFFFF
        elif op == OP_CALL:
            self._ret_stack.append(self.pc & 0xFFFF)
            self.pc = imm & 0xFFFF
        elif op == OP_RET:
            if self._ret_stack:
                self.pc = self._ret_stack.pop() & 0xFFFF
        elif op == OP_LDIO:
            self.regs[0] = self._read_byte(mmio_addr(imm))
        elif op == OP_STIO:
            self.bus.write_cpu(mmio_addr(imm), self.regs[0])

    def step(self) -> None:
        self.fetch_decode_execute()

    def sync_from_micro(self, regs: list[int], flag_z: bool, pc: int) -> None:
        self.regs = list(regs)
        self.flag_z = flag_z
        self.pc = pc

File: crates/plover_core/src/fast.rs
use crate::alu::{apply_add, apply_beq_compare, apply_cmp_flags, alu8};
use crate::alu16::{add16, cmp16_u};
use crate::isa::*;
use plover_mmu::{mmio_addr, MemoryBus};

pub struct MacroFastPath {
    pub bus: MemoryBus,
    pub pc: u16,
    pub regs: [u8; 4],
    pub regs16: [u16; 4],
    pub flag_z: bool,
    pub flag_c: bool,
    pub halted: bool,
    pub ret_stack: Vec<u16>,
}

impl MacroFastPath {
    pub fn new(bus: MemoryBus) -> Self {
        Self {
            bus,
            pc: 0,
            regs: [0; 4],
            regs16: [0; 4],
            flag_z: false,
            flag_c: false,
            halted: false,
            ret_stack: Vec::new(),
        }
    }

    fn read_byte(&self, addr: u16) -> u8 {
        self.bus.read_cpu(addr)
    }

    pub fn fetch_decode_execute(&mut self) {
        if self.halted {
            return;
        }
        let fa = self.bus.fetch_addr(self.pc);
        let op = self.read_byte(fa);

        let (imm, wide_imm16): (u16, bool) = if is_wide_imm16(op) {
            let v = u16::from(self.read_byte(fa.wrapping_add(1)))
                | (u16::from(self.read_byte(fa.wrapping_add(2))) << 8);
            self.pc = fa.wrapping_add(3);
            (v, true)
        } else if is_wide_abs16(op) {
            let v = u16::from(self.read_byte(fa.wrapping_add(1)))
                | (u16::from(self.read_byte(fa.wrapping_add(2))) << 8);
            self.pc = fa.wrapping_add(3);
            (v, false)
        } else if matches!(op, OP_RET | OP_HALT | OP_ADD_RR) {
            self.pc = fa.wrapping_add(1);
            (0, false)
        } else {
            let v = u16::from(self.read_byte(fa.wrapping_add(1)));
            self.pc = fa.wrapping_add(2);
            (v, false)
        };

        match op {
            OP_HALT => self.halted = true,
            OP_ADD => {
                let (r, z, c) = apply_add(self.regs, imm as u8);
                self.regs = r;
                self.flag_z = z;
                self.flag_c = c;
            }
            OP_ADD_RR => {
                let res = alu8(self.regs[0], self.regs[1], 1);
                self.regs[2] = res.y;
                self.flag_z = res.zero;
                self.flag_c = res.cout;
            }
            OP_MOV => {
                let dst = ((imm >> 4) & 3) as usize;
                let src = (imm & 3) as usize;
                self.regs[dst] = self.regs[src];
            }
            OP_CMP => {
                let (z, c) = apply_cmp_flags(self.regs, imm as u8);
                self.flag_z = z;
                self.flag_c = c;
            }
            OP_WADD_RR => {
                let res = add16(self.regs16[0], self.regs16[1]);
                self.regs16[2] = res.y;
                self.flag_z = res.zero;
                self.flag_c = res.cout;
            }
            OP_WMOV => {
                let dst = ((imm >> 4) & 3) as usize;
                let src = (imm & 3) as usize;
                self.regs16[dst] = self.regs16[src];
            }
            OP_WCMP16 if wide_imm16 => {
                let res = cmp16_u(self.regs16[0], imm);
                self.flag_z = res.zero;
                self.flag_c = res.cout;
            }
            OP_BCS => {
                if self.flag_c {
                    self.pc = imm;
                }
            }
            OP_LDA => self.regs[0] = self.read_byte(imm as u16),
            OP_STA | OP_STA16 => self.bus.write_cpu(imm as u16, self.regs[0]),
            OP_BEQ => {
                let (z, c) = apply_beq_compare(self.regs, imm as u8);
                self.flag_z = z;
                self.flag_c = c;
                if self.flag_z {
                    self.pc = imm;
                }
            }
            OP_JMP => self.pc = imm,
            OP_CALL => {
                self.ret_stack.push(self.pc);
                self.pc = imm;
            }
            OP_RET => {
                if let Some(addr) = self.ret_stack.pop() {
                    self.pc = addr;
                }
            }
            OP_LDIO => self.regs[0] = self.read_byte(mmio_addr(imm as u8)),
            OP_STIO => self.bus.write_cpu(mmio_addr(imm as u8), self.regs[0]),
            _ => {}
        }
    }

    pub fn step(&mut self) {
        self.fetch_decode_execute();
    }

    pub fn run(&mut self, max_steps: usize) {
        for _ in 0..max_steps {
            if self.halted {
                break;
            }
            self.step();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use plover_mmu::MB_STATUS;

    #[test]
    fn ldio_stio_parity() {
        let mut bus = MemoryBus::default();
        bus.map_mode = 1;
        let mut fast = MacroFastPath::new(bus);
        fast.regs[0] = 0xAB;
        fast.pc = 0;
        let prog = [OP_LDIO, 0x00, OP_STIO, 0x04, OP_HALT];
        for (i, &b) in prog.iter().enumerate() {
            fast.bus.ram.write(i as u16, b);
        }
        fast.run(10);
        assert_eq!(fast.regs[0], 0x08); // APU_READY on idle status
        assert_eq!(fast.bus.mailbox.read(MB_STATUS), 0x08);
    }
}
